src/v/storage/segment.cc

```diff
@@ -202,15 +203,35 @@ ss::future<> segment::release_appender() {
                      a = std::move(a),
                      c = std::move(c),
                      i = std::move(i)]() mutable {
-                        return write_lock().then(
```

Comment:
I really don't understand the deadlock scenario described in the commit message. could you provide some more detail about the scenario? (hint: it's hard to know what `we` is, for instance, in `...when we want to create reader...`). is this related to the reader cache or a totally separate bug? I have a hunch that this particular problem can be solved without the repeat loop, but I don't fully understand the scenario yet.

Reply:
I've added this as a commend and commit message: 
Segment close operation is destructive this makes it to require segment write 
lock. When any of the readers is open for given segment it can not be closed 
without waiting since each reader holds a read lock for the segment. In order to
be able to roll the segment without waiting for all readers using this segment 
to be closed when immediate write lock is impossible we dispatch segment close
operation in background. Background close operation acquires a write lock and 
then closes the segment. 

Seastar `read_write_lock` is backed by the semaphore. Acquiring read lock 
requires at least one unit to be available in the semaphore while acquiring 
write lock requires all of the units to be available. 
(semaphore is initialized with max integer count of units). All of the wait 
requests are serialized into semaphore `_waiters` list. This cause the situation
in which any read lock can not be acquired if write lock request is in the 
`_waiters` list.

Example
```
┌──────┬───────┬──────┬──────┬──────┐
│ read │ write │ read │ read │ read │
└──────┴───────┴──────┴──────┴──────┘
   0       1      2      3      4
```

In this situation after read lock from position 0 is acquired any further read 
locks can not be acquired since write lock request is 'blocking' `_waiters` list.

Both of described rules together may lead to deadlock situation in following 
situation:

1. create reader for segment s1.
2. append entries to log to cause segment roll 
3. create reader for segments s1 & s2. 

In this situation step `3.` will never succeed until reader from step 1. is alive.

Using spin lock approach allow us to wait for the write lock without inserting 
wait request to underlying semaphore `_waiters` list.

Reply:
Thanks for this comment, it is really amazing!

The flip side of this seems to be that reader can deny the ability for storage to roll logs as they grow, right? How problematic might that be if there are long lived active readers and producers? If that's problematic then it seems that we need a way to guarantee progress on rolling.

Given all the work in this patch series that allows log reader instances to be suspended / reconfigured, would it make sense to have a policy by which readers released their locks when they went back into the reader cache? I think there are some other tricks I think we could do that doesn't have the starvation property.